### DISCUSSION TITLE
Networking 24 Siemens PLCs and controlling them remotely

### DISCUSSION THREAD

**Discussion by assenzostefano** [pymodbus-dev/pymodbus#2114] [General]
Hello, my name is Stefano and I am 17 years old. I'm learning how to use PLCs (more specifically Siemens LOGO PLCs!) at school, at the moment to upload our LAD file, we have to pass the file to the professor, he has to pass the file to his laptop, get close to the PLC and connect with the Ethernet and finally upload the program.
This method takes a lot of time, especially when the professor is busy with something else, so we thought of networking all 24 PLCs via NAT, then creating a back-end in Python that manages the PLCs, so we students can upload the LAD file directly from our PC even if we are not physically connected.

I discovered this library during my research, it has a lot of functionality but I didn't find the one thing I needed, which is to upload a LAD file via Logo Soft Comfort (with the Dynamic DNS function) but instead of putting the NAT IP and the port of that specific PLC, it uses a single IP (i.e. the back-end IP) that decides which PC to assign a specific PLC to.
Now you may be wondering, "how do you decide which PLC to give to each PC?", the user simply has to access a web page, select the PLC to which they want to connect and enter the back-end IP, hence the back-end saves the selection and when uploading the LAD file via Logo Soft Comfort sees which PLC you want to connect to, and voilà, the file will have been uploaded to the PLC.

The only problem is that I don't know how the PLC communicates, i.e. whether I just use the Python library "socket" and activate the communication between PC - Back-end - PLC or whether the PLC uses a different method for communication.

Thank you very much in advance

**Comment by janiversen** [comment]
I am not sure the lad file is uploaded via the modbus connection, at least siemens prof. PLC have a separate configuration channel.

But assuming it uploads via modbus then you need to use the file transfer requests.

Your idea is very good, and very doable, it just need you to

- first dive into the device manual and see how the lad file is uploaded 
- secondly make a program that does the upload using a library
- make the web etc.

**Reply by assenzostefano** [reply]
Thank you very much for answering me, these days I try to see the instruction manuals provided by Siemens. As soon as I discover how Logo Soft Comfort communicates with the PLC, if you want I will add the whole explanation in this thread.

**Reply by janiversen** [reply]
Not really needed.

**Reply by janiversen** [reply]
But thanks for the offer.
